io/src/main/java/io/smallrye/common/io/PathUtils.java

```diff
@@ -0,0 +1,2139 @@
+import static io.smallrye.common.constraint.Assert.checkNotNullParam;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.nio.file.AccessMode;
+import java.nio.file.CopyOption;
+import java.nio.file.DirectoryIteratorException;
+import java.nio.file.DirectoryNotEmptyException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.FileSystemException;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.NotDirectoryException;
+import java.nio.file.NotLinkException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SecureDirectoryStream;
+import java.nio.file.attribute.BasicFileAttributeView;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.FileTime;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+
+import io.smallrye.common.constraint.Assert;
+
+/**
+ * Utilities for {@link java.nio.file.Files}. Fluent style methods, more input
+ * types and unchecked exceptions make these more convenient.
+ *
+ * @author Boris Unckel
+ *
+ */
+public final class PathUtils {
+
+    public static final LinkOption[] NO_FOLLOW_LINK_OPTION_ARRAY = { LinkOption.NOFOLLOW_LINKS };
+
+    /**
+     * A wrapper around {@link java.nio.file.Paths#get(String, String...)} with
+     * explicit null check.
+     *
+     * @param first part of the path, must be not null, not empty and a valid path
+     * @param more optional parts of the path
+     * @return the path constructed out of first and more
+     * @throws IllegalArgumentException if first is null or empty
+     */
+    public static Path get(final String first, final String... more) {
+        return Paths.get(checkNotEmptyAfterTrimParam("first", first), more);
+    }
+
+    /**
+     * A wrapper around {@link java.nio.file.Paths#get(URI)} with explicit null
+     * check.
+     *
+     * @param uri for the path, must be not null and a valid path
+     * @return the path constructed out of the uri
+     * @throws IllegalArgumentException if uri is null
+     */
+    public static Path get(final URI uri) {
+        return Paths.get(checkNotNullParam("uri", uri));
+    }
+
+    /**
+     * A wrapper around {@link java.nio.file.Files#newDirectoryStream(Path)} with
+     * explicit null check and unchecked {@code java.io.UncheckedIOException}.
+     *
+     * @param dir to the directory to list, must be not null.
+     * @return the elements in the directory as DirectoryStream
+     * @throws IllegalArgumentException if the path is {@code null}
+     * @throws UncheckedIOException around original IOExceptions
+     */
+    public static DirectoryStream<Path> newDirectoryStream(final Path dir) throws UncheckedIOException {
```

Comment:
I personally don't much like wrapping with `UncheckedIOException`.  We don't do that in just about any other place that I can think of. How would you justify this feature?

Reply:
Before I started with any of the coding I have searched for places which might be worth to be refactored from classic java.io to java.nio. Many follow a pattern with private utility methods declaring a throws IOException (without Exception handling) and a major place where no Exception handling logic appears but to rethrow it as unchecked Exception.
You're right that currently the usage is different. I think the basic reason is backward compatibility of the JDK and missing 3rd party libraries supporting it.

I have added a section for this in the class javadoc.

Reply:
I think the feature would *ideally* have to be justifiable within the context of SmallRye projects, or else be widely necessary by consumers of SmallRye projects, which is why I asked.  Unchecked versions of things is a specific enough need that I would imagine it to be framed specifically that way: e.g. `UncheckedPaths`, `UncheckedFiles` etc.  Also, specifically for methods like this one which return an object which is a subtype of `Closeable`, I'd expect a wrapper object which throws an unchecked exception on close e.g. `interface UncheckedDirectoryStream<Path> extends DirectoryStream<Path> { void close() throws UncheckedIOException; }`.

Put more simply, making unchecked versions of NIO methods is a separate concern from formalizing `null` checks according to a different scheme (though it would certainly be reasonable to perform `null` checks according to that scheme from such an API), and I'd expect the API to reflect that.

Reply:
Yes, something I didn't think of: It's not enough to provide "unchecked" utility methods, the underlying objects should behave in the same way.
This leads to a new problem: If UncheckedPath extends Path it's possible to provide different exception signatures, but a wrapper implementation can only throw either checked xor unchecked exceptions, not depending on a cast. All other objects (channels, buffer...) would still require checked ex behaviour.

In short: this is not ready to launch, neither in total nor in parts.